// mtproto-shared/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _dtime = require('./dtime');

Object.defineProperty(exports, 'dTime', {
  enumerable: true,
  get: function () {
    return _dtime.dTime;
  }
});
Object.defineProperty(exports, 'dTimePure', {
  enumerable: true,
  get: function () {
    return _dtime.dTimePure;
  }
});

var _memo = require('./memo');

Object.defineProperty(exports, 'memoize', {
  enumerable: true,
  get: function () {
    return _memo.memoize;
  }
});

var _timeout = require('./timeout');

Object.defineProperty(exports, 'pause', {
  enumerable: true,
  get: function () {
    return _timeout.pause;
  }
});
Object.defineProperty(exports, 'smartTimeout', {
  enumerable: true,
  get: function () {
    return _timeout.smartTimeout;
  }
});
Object.defineProperty(exports, 'immediate', {
  enumerable: true,
  get: function () {
    return _timeout.immediate;
  }
});
Object.defineProperty(exports, 'immediateWrap', {
  enumerable: true,
  get: function () {
    return _timeout.immediateWrap;
  }
});
