// telegram-mtproto/lib/service/authorizer/rsa-keys-manger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KeyManager = KeyManager;

require('../main/index.h');

require('../api-manager/index.h');

var _tl = require('../../tl');

var _writer = require('../../tl/writer');

var _bin = require('../../bin');

function KeyManager(uid, publisKeysHex, publicKeysParsed) {
  var prepared = false;

  function mapPrepare({ modulus, exponent }) {
    var RSAPublicKey = new _tl.Serialization({}, uid);
    var rsaBox = RSAPublicKey.writer;
    (0, _writer.writeBytes)(rsaBox, (0, _bin.bytesFromHex)(modulus) /*, 'n'*/);
    (0, _writer.writeBytes)(rsaBox, (0, _bin.bytesFromHex)(exponent) /*, 'e'*/);

    var buffer = rsaBox.getBuffer();

    var fingerprintBytes = (0, _bin.sha1BytesSync)(buffer).slice(-8);
    fingerprintBytes.reverse();
    var key = (0, _bin.bytesToHex)(fingerprintBytes);
    //$FlowIssue obj[number] === obj[stringNumber]
    publicKeysParsed[key] = {
      modulus,
      exponent
    };
  }

  function prepareRsaKeys() {
    if (prepared) return;
    publisKeysHex.forEach(mapPrepare);

    prepared = true;
  }

  function selectRsaKeyByFingerPrint(fingerprints) {
    prepareRsaKeys();

    var fingerprintHex = void 0,
        foundKey = void 0;
    for (var _iterator = fingerprints, _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : _iterator[Symbol.iterator]();;) {
      var _ref;

      if (_isArray) {
        if (_i >= _iterator.length) break;
        _ref = _iterator[_i++];
      } else {
        _i = _iterator.next();
        if (_i.done) break;
        _ref = _i.value;
      }

      var fingerprint = _ref;

      fingerprintHex = (0, _bin.strDecToHex)(fingerprint);
      //$FlowIssue obj[number] === obj[stringNumber]
      foundKey = publicKeysParsed[fingerprintHex];
      if (foundKey) return Object.assign({ fingerprint }, foundKey);
    }
    throw new Error('[Key manager] No public key found');
  }

  return {
    prepare: prepareRsaKeys,
    select: selectRsaKeyByFingerPrint
  };
}

exports.default = KeyManager;
