// telegram-mtproto/lib/util/either.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Either = undefined;
exports.of = of;
exports.Right = Right;
exports.Left = Left;
exports.fromNullable = fromNullable;
exports.withPred = withPred;

var _result = require('folktale/result');

var Result = _interopRequireWildcard(_result);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

class Either {
  constructor(inner) {
    Object.defineProperty(this, 'inner', {
      value: inner,
      enumerable: false
    });
    this.chiral = this.matchWith({
      Right: () => 'Right',
      Left: () => 'Left'
    });
    this.value = inner.value;
  }
  map(fn) {
    return new Either(this.inner.map(fn));
  }
  mapError(fn) {
    return new Either(this.inner.mapError(fn));
  }
  merge() {
    return this.inner.merge();
  }
  chain(fn) {
    //$FlowIssue
    var chained = this.matchWith({
      Right: res => fn(res.value),
      Left: () => this
    });
    return chained;
  }
  getOrElse(defaults) {
    return this.inner.getOrElse(defaults);
  }
  orElse(fn) {
    //$FlowIssue
    var chained = this.matchWith({
      Right: () => this,
      Left: res => fn(res.value)
    });
    return chained;
  }
  matchWith(matcher) {
    return this.inner.matchWith({
      Ok: matcher.Right,
      Error: matcher.Left
    });
  }
  cata(matcher) {
    return this.matchWith(matcher);
  }
}

exports.Either = Either;
function of(obj) {
  var inner = Result.of(obj);
  return new Either(inner);
}

function Right(obj) {
  return of(obj);
}

function Left(obj) {
  var inner = Result.Error(obj);
  return new Either(inner);
}

function fromNullable(obj) {
  return new Either(Result.fromNullable(obj));
}

function withPred(pred, val) {
  if (pred(val)) {
    return Right(val);
  } else {
    return Left(val);
  }
}

exports.default = of;
