// telegram-mtproto/lib/dc-status/seq-no.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _ramda = require('ramda');

class SeqNo {
  constructor() {
    this.limit = 100;

    this.init();
  }
  init() {
    this.count = 0;
    this.first = 0;
    this.current = 0;
    this.holes = 0;
    this.history = [];
    this.uniq = [];
    this.hasHoles = false;
  }
  getNext(notContentRelated) {
    var seqNo = this.count * 2;
    if (!notContentRelated) {
      seqNo++;
      this.count++;
    }
    this.push(seqNo);
    return seqNo;
  }
  push(next) {
    this.current = Math.max(this.current, next);
    this.history = (0, _ramda.append)(next, this.history);
    if (this.history.length >= this.limit) {
      var dropCount = this.history.length - this.limit;
      this.history = (0, _ramda.drop)(dropCount, this.history);
    }
    this.uniq = (0, _ramda.uniq)(this.history);
    this.findMin();
    this.countHoles();
  }
  findMin() {
    var ln = this.uniq.length;
    var min = Infinity;
    for (var i = 0; i < ln; i++) {
      var val = this.uniq[i];
      if (val < min) min = val;
    }
    this.first = min;
  }
  countHoles() {
    var ln = this.uniq.length;
    if (ln === 0) {
      this.holes = 0;
      return;
    }
    var range = this.current - this.first;
    this.holes = range - ln;
    this.hasHoles = this.holes === 0;
  }
}
exports.default = SeqNo;
