// telegram-mtproto/lib/state/signal/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.networker = exports.uid = exports.homeDc = exports.whenActive = exports.active = exports.rootSignal = undefined;

var _core = require('../core');

var _core2 = _interopRequireDefault(_core);

require('../index.h');

var _mostSubject = require('most-subject');

require('most');

var _configProvider = require('../../config-provider');

var _configProvider2 = _interopRequireDefault(_configProvider);

var _pullStream = require('../../pull-stream');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var rootSignal = exports.rootSignal = (0, _mostSubject.async)();
var root = rootSignal.multicast();

var isActive = root.map(state => state.active).skipRepeats();

exports.active = isActive;
var whenActive = exports.whenActive = (0, _pullStream.faucetC)(isActive);

var homeDc = exports.homeDc = root.map(state => state.homeDc).skipRepeats();

var uid = exports.uid = root.map(state => state.uid).skipRepeats();

var networker = exports.networker = root.map(state => state.networker);

var stateModel = {
  active: isActive,
  homeDc,
  uid,
  networker
};

exports.default = stateModel;
