// telegram-mtproto/lib/state/action/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.API = exports.NETWORKER_STATE = exports.AUTH = exports.NET = exports.MAIN = undefined;

require('axios');

var _netMessage = require('../../service/networker/net-message');

var _networker = require('../../service/networker');

var _request = require('../../service/main/request');

var _request2 = _interopRequireDefault(_request);

var _helpers = require('../helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MAIN = exports.MAIN = {
  TEST: (0, _helpers.doubleCreator)('test'),
  INIT: (0, _helpers.doubleCreator)('init'),
  SWITCH_ON: (0, _helpers.doubleCreator)('module loaded'),
  ACTIVATED: (0, _helpers.doubleCreator)('instance activated')
};

var networkerMeta = (_, dc) => ({ _: 'networker', id: dc });
var apiMeta = (_, id) => ({ _: 'api', id });
var NET = exports.NET = {
  SEND: (0, _helpers.doubleCreator)('net/send', networkerMeta),
  RECEIVE_RESPONSE: (0, _helpers.doubleCreator)('net/response'),
  NETWORK_ERROR: (0, _helpers.doubleCreator)('net/error')
};

var AUTH = exports.AUTH = {
  SET_AUTH_KEY: (0, _helpers.doubleCreator)('auth/auth_key set', networkerMeta),
  SET_SERVER_SALT: (0, _helpers.doubleCreator)('auth/server_salt set', networkerMeta),
  SET_SESSION_ID: (0, _helpers.doubleCreator)('auth/session_id set', networkerMeta)
};

var NETWORKER_STATE = exports.NETWORKER_STATE = {
  RESEND: {
    ADD: (0, _helpers.doubleCreator)('networker/resend add', networkerMeta),
    DEL: (0, _helpers.doubleCreator)('networker/resend delete', networkerMeta)
  },
  SENT: {
    ADD: (0, _helpers.doubleCreator)('networker/sent add', networkerMeta),
    DEL: (0, _helpers.doubleCreator)('networker/sent delete', networkerMeta)
  },
  PENDING: {
    ADD: (0, _helpers.doubleCreator)('networker/pending add', networkerMeta),
    DEL: (0, _helpers.doubleCreator)('networker/pending delete', networkerMeta)
  }
};

var API = exports.API = {
  NEW_REQUEST: (0, _helpers.doubleCreator)('api/request new', apiMeta),
  DONE_REQUEST: (0, _helpers.doubleCreator)('api/request done', apiMeta),
  CALL_TASK: (0, _helpers.doubleCreator)('api/call-task new'),
  CALL_RESULT: (0, _helpers.doubleCreator)('api/call-task done')
};
